sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -17,61 +17,7 @@ import scalaz.syntax.semigroup._
 import java.nio.{ByteBuffer, ByteOrder}
 
 /** Values */
-sealed abstract class Value extends CidContainer[Value] with Product with Serializable {
-  import Value._
-
-  final override protected def self: this.type = this
-
-  final override def mapCid(f: ContractId => ContractId): Value = {
-    def go(v0: Value): Value =
-      // TODO (FM) make this tail recursive
-      v0 match {
-        case ValueContractId(coid) => ValueContractId(f(coid))
-        case ValueRecord(id, fs) =>
-          ValueRecord(
-            id,
-            fs.map({ case (lbl, value) =>
-              (lbl, go(value))
-            }),
-          )
-        case ValueVariant(id, variant, value) =>
-          ValueVariant(id, variant, go(value))
-        case x: ValueCidlessLeaf => x
-        case ValueList(vs) =>
-          ValueList(vs.map(go))
-        case ValueOptional(x) => ValueOptional(x.map(go))
-        case ValueTextMap(x) => ValueTextMap(x.mapValue(go))
-        case ValueGenMap(entries) =>
-          ValueGenMap(entries.map { case (k, v) => go(k) -> go(v) })
-      }
-    go(this)
-  }
```

Comment:
Moved this method into the various subclasses so that the types can be refined.

Reply:
Well the returned value does NOT inhabit `this.type` because `this.type` is just a singleton type. I agree that we should have something like `this.classType` in Scala that allows you to return any instance of the same class as `this`, but that's probably incompatible with the rest of the Scala type system.